### DISCUSSION TITLE
useGLTF typescript access geometry in typesafe way

### DISCUSSION THREAD

**Discussion by Dave-Patsy** [pmndrs/drei#1337] [Q&A] [answered]
I am looking at the Shoe configurator example and trying to refactor it to typescript

This is the code in the example. However, Property 'nodes' and 'materials ' does not exist on type 'GLTF'
it does work, but  'nodes' and 'materials ' are type any
```
const { nodes,  materials }= useGLTF("/models/shoe-draco.glb")
```

these are the variables of useGLTF
```
const { parser,asset,scenes,scene,userData,animations,cameras} = useGLTF("/models/shoe-draco.glb")
```

this will give me the geometries for all the meshes, but it is not typesafe.
```
scene.children.map(x=>x.children.map(y=>y.geometry))

```
I am trying to loop through the model to create multiple meshes for example.

```
return(
    <group
      ref={ref}
      dispose={null}
      onPointerOver={(e) => (e.stopPropagation(), set(e.object.material.name))}
      onPointerOut={(e) => e.intersections.length === 0 && set(null)}
      onPointerMissed={() => (state.current = null)}
      onClick={(e) => (e.stopPropagation(), (state.current = e.object.material.name))}>
      {scene.children.map(x=>x.children.map(y=>{
        <mesh 
           receiveShadow 
           castShadow 
           geometry={y.geometry} 
           material={y.materials} 
           material-color={"#ffffff"}>
      }
    </group>
)
```
I tried
```

 if(typeof y === THREE.Mesh){
  y.geometry
}
```
But then, the geometries became undefined.

Is there a typesafe way to access the geometries and materials of an Object3D?

**Comment by myznikovgleb** [comment, marked as answer]
@Dave-Patsy 

I guess you just miss the point of usage [gtlfjsx](https://github.com/pmndrs/gltfjsx) or its [web version](https://gltf.pmnd.rs/) to work with GLTF-models. This command-line tool also support type safe user experience

I also try to explain what happens in this code snippet

```
const { nodes,  materials }= useGLTF("/models/shoe-draco.glb")
```

`useGLTF` return value with `GLTF` type imported from [three-stdlib](https://github.com/pmndrs/three-stdlib/blob/main/src/loaders/GLTFLoader.d.ts). You can see that `GLTF` type does not contain `nodes` or `materials` fields.  Why? I guess that it is simply imposable for third-party library (as three-stdlib) to predict what types of nodes (i.e. meshes) and materials will be stored at your brand new model

And here you have to options. The first one is to manually describe content of your model with corresponding types and names. But I do not recommend to do that. The last one is to use convenient solution provided by developers of this library that I've mentioned above
